Sources/SwiftInspectorCommands/Tests/TypeLocationCommandSpec.swift

```diff
@@ -0,0 +1,189 @@
+//
+// Distributed under the MIT License
+//
+// Permission is hereby granted, free of charge, to any person obtaining a copy
+// of this software and associated documentation files (the "Software"), to deal
+// in the Software without restriction, including without limitation the rights
+// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in all
+// copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+// SOFTWARE.
+
+import Nimble
+import Quick
+import Foundation
+
+@testable import SwiftInspectorCommands
+@testable import SwiftInspectorCore
+
+final class TypeLocationCommandSpec: QuickSpec {
+
+  override func spec() {
+    var pathURL: URL!
+
+    describe("TypeLocationCommand") {
+      afterEach {
+        guard let pathURL = pathURL else {
+          return
+        }
+        try? Temporary.removeItem(at: pathURL)
+      }
+
+      describe("run") {
+
+        context("with no arguments") {
+          it("fails") {
+            let result = try? TestTypeLocationTask.run(path: nil, name: nil)
+            expect(result?.didFail) == true
+          }
+        }
+
+        context("path is valid") {
+          beforeEach { pathURL = try? Temporary.makeFile(content: "struct Foo { }") }
+
+          context("with no --name argument") {
+            it("fails") {
+              let result = try? TestTypeLocationTask.run(path: pathURL.path, name: nil)
+              expect(result?.didFail) == true
+            }
+          }
+
+          context("with an empty --name argument") {
+            it("fails") {
+              let result = try? TestTypeLocationTask.run(path: pathURL.path, name: "")
+              expect(result?.didFail) == true
+            }
+          }
+        }
+
+        context("name is valid") {
+          context("with no --path argument") {
+            it("fails") {
+              let result = try? TestTypeLocationTask.run(path: nil, name: "Foo")
+              expect(result?.didFail) == true
+            }
+          }
+
+          context("with an empty --path argument") {
+            it("fails") {
+              let result = try? TestTypeLocationTask.run(path: "", name: "Foo")
+              expect(result?.didFail) == true
+            }
+          }
+        }
+
+        context("when path is a directory") {
```

Comment:
I generally think SwiftInspector should be moving this direction and pushing the iteration logic up a level.
I think this keep SwiftInspector focused.

When you have directories you also need to start thinking of how to handle edge cases (e.g. same type, multiple modules). You get a bit of that in a file with nested types, but it's less extreme. I think it's better if those edge cases are pushed to consumers.

Reply:
There's definitively a tradeoff between making this easier to use and making this easier to maintain. I think most of the use cases for SwiftInspector are actually going through folders, so I'd prefer to allow folders to be traversed for the sake of usability of the tool even if that means maintenance gets more complicated. For this case where we're actually just trying to discover where a specific type is declared I think limiting to the file makes sense though. For something like the imports command that wouldn't make as much sense.

Reply:
@fdiaz I see how it's useful for some commands. When used at scale, I can also see performance wins. Staying in Swift is more performance than calling into Swift many times from a higher level language like Ruby or Bash. But at the same time, you could solve that problem by passing a file list....

As this tool becomes more mature, I think we'll want to standardize more. But if you're OK with it (which it sounds like you are by this comment) I guess we can have commands handle this differently for now.

Maybe this is one of those things where the answer will become more obvious over time.

Reply:
Another thing to consider is _how_ we intend for this to be used. At work, we usually invoke this from a higher level language and no human is running it on the CLI. If you are actually running this yourself on the CLI I think it becomes harder to reason about directories. Not sure exactly what to do with that thought, but something to chew on.

Reply:
> If you are actually running this yourself on the CLI I think it becomes harder to reason about directories

I feel the opposite. If we don't provide a way to easily run in directories, this tool becomes a lot less useful since you have to manually traverse the directories and run this file on each Swift file. That's not as useful as inspecting the usage of code in your codebase.

Reply:
Ah let me clarify. I think we're actually mostly in agreement.

What I meant is that dealing with the _output_ can be hard when you're using this tool manually on the CLI. The reason is that the file paths are long and then the output is hard to visually read. Also, as we discussed, the potential for spaces in the file path makes it hard to use tools like `cut`. This last issue could be mitigated by putting the file path last, or in other ways.

I thought about this some more today. The more I think about it I do see value in supporting traversing folders for as many commands as possible _eventually_. I think the thing we'd need to figure out is how best to display the paths.

I kind of like how `grep` works in this regard. By default `grep` accepts a path.

```shell
grep "myword" ./Foo/Swift
```

But you can also pass `-R` to traverse directories.

```shell
grep "myword" -R .
```

Maybe the file names aren't as big of an issue as I thought. I guess with `grep` I usually pipe to a file and then process the output from there 🤔  It seems reasonable that I would do the same for SwiftInspector output. We could solve the spaces in paths issue in any number of ways, including by adding a way to specify an output separator, which you could set to a tab instead of a space.

Anyway, I think that there is an opportunity to _standardize_ on an approach for this across commands _eventually_.

Honestly though I think a more pressing issue is getting these commands to work with relative paths 😂

Reply:
@fdiaz I think we landed in a good place. I'm resolving this conversation. If you think we still have more to discuss, please unresolve or if that's not possible (I forget if it is) add a new comment 👍